File: python/singleton/singleton.py
class Pool:
    _instance = None

    def __init__(self):
        self.i = 0

    @staticmethod
    def get_instance():
        if Pool._instance is None:
            Pool._instance = Pool()
        return Pool._instance

    def inc(self):
        self.i += 1

    def __str__(self):
        return f"Current value: {str(self.i)}"


if __name__ == '__main__':
    Pool.get_instance().inc()
    print(Pool.get_instance())
    Pool.get_instance().inc()
    Pool.get_instance().inc()
    print(Pool.get_instance())

File: cpp/singleton/singleton.cpp
#include <iostream>

class Pool {
public:
    static Pool &get_instance() {
        static Pool p;
        return p;
    }

    Pool(const Pool &pool) = delete;

    Pool &operator=(const Pool &pool) = delete;

    void inc() {
        _i += 1;
    }

    int get_i() const {
        return _i;
    }

private:
    Pool() = default;

    int _i = 0;
};

int main() {
    std::cout << Pool::get_instance().get_i() << std::endl;
    Pool::get_instance().inc();
    Pool::get_instance().inc();
    std::cout << Pool::get_instance().get_i() << std::endl;
    return 0;
}
